File: client/features/learning/api/revision.ts
import { apiFetch } from "@/lib/api";

export type RevisionPlan = {
  id: string;
  student_profile_id: string;
  subject: string;
  exam: string;
  window_start: string;
  window_end: string;
  status: "DRAFT" | "ACTIVE" | "ARCHIVED";
  created_at: string;
  updated_at: string;
};

export type RevisionTask = {
  id: string;
  plan_id: string;
  topic: string;
  window_start: string;
  window_end: string;
  status: "PENDING" | "DONE" | "SKIPPED";
  priority: 1 | 2 | 3;
  source: "seeded" | "rebalanced";
  completed_at?: string;
  created_at: string;
};

export async function listRevisionPlans(studentProfileId?: string): Promise<RevisionPlan[]> {
  const q = studentProfileId ? `?student_profile_id=${studentProfileId}` : "";
  const res = await apiFetch<RevisionPlan[]>(`/me/revision-plans${q}`);
  return res.data ?? [];
}

export async function createRevisionPlan(input: {
  subject: string;
  exam?: string;
  window_start: string; // YYYY-MM-DD
  window_end: string;
  studentProfileId?: string;
}): Promise<RevisionPlan> {
  const q = input.studentProfileId ? `?student_profile_id=${input.studentProfileId}` : "";
  const res = await apiFetch<RevisionPlan>(`/me/revision-plans${q}`, {
    method: "POST",
    body: JSON.stringify({
      subject: input.subject,
      exam: input.exam ?? "",
      window_start: input.window_start,
      window_end: input.window_end,
    }),
  });
  return res.data;
}

export async function rebalancePlan(planId: string): Promise<number> {
  const res = await apiFetch<{ reprioritised: number }>(
    `/me/revision-plans/${planId}/rebalance`,
    { method: "POST" },
  );
  return res.data?.reprioritised ?? 0;
}

export async function completeTask(taskId: string): Promise<void> {
  await apiFetch(`/me/revision-plans/tasks/${taskId}/complete`, { method: "POST" });
}

File: internal/domain/revision/revision.go
// Package revision — adaptive revision planner (feature 1, migration 000075).
// A plan schedules a learner's weak topics across a study window; a weekly
// rebalance promotes weak units and compresses mastered ones based on real
// mastery data. No exam predictions — only the learner's own attempt history.
package revision

import (
	"context"
	"time"

	"github.com/google/uuid"
)

// Plan — one revision plan per (learner, subject, exam) with a study window.
type Plan struct {
	ID               uuid.UUID `json:"id"`
	StudentProfileID uuid.UUID `json:"student_profile_id"`
	Subject          string    `json:"subject"`
	Exam             string    `json:"exam"`
	WindowStart      time.Time `json:"window_start"`
	WindowEnd        time.Time `json:"window_end"`
	Status           string    `json:"status"` // DRAFT|ACTIVE|ARCHIVED
	CreatedAt        time.Time `json:"created_at"`
	UpdatedAt        time.Time `json:"updated_at"`
}

// Task — one unit/topic scheduled inside a plan's window.
type Task struct {
	ID          uuid.UUID  `json:"id"`
	PlanID      uuid.UUID  `json:"plan_id"`
	Topic       string     `json:"topic"`
	WindowStart time.Time  `json:"window_start"`
	WindowEnd   time.Time  `json:"window_end"`
	Status      string     `json:"status"`   // PENDING|DONE|SKIPPED
	Priority    int        `json:"priority"` // 1 (high) .. 3 (low)
	Source      string     `json:"source"`   // seeded|rebalanced
	CompletedAt *time.Time `json:"completed_at,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
}

// Repository — persistence for revision plans + tasks.
type Repository interface {
	CreatePlan(ctx context.Context, p *Plan) error
	GetPlan(ctx context.Context, id uuid.UUID) (*Plan, error)
	ListPlans(ctx context.Context, studentProfileID uuid.UUID) ([]Plan, error)
	AddTasks(ctx context.Context, tasks []Task) error
	ListTasks(ctx context.Context, planID uuid.UUID) ([]Task, error)
	// SetTaskPriority reprioritises a task and stamps its source (rebalance).
	SetTaskPriority(ctx context.Context, taskID uuid.UUID, priority int, source string) error
	CompleteTask(ctx context.Context, taskID uuid.UUID) error
}
